File: src/utils/urlMetadata/twitterParser.py
import re
import requests
import urllib.parse
from ..logger import logger

TWITTER_DOMAINS = ['twitter.com', 'x.com', 'fxtwitter.com', 'fixupx.com', 'vxtwitter.com']

def match(url):
    try:
        parsed = urllib.parse.urlparse(url)
        hostname = parsed.hostname
        if not hostname:
            return False
        return hostname in TWITTER_DOMAINS or any(hostname.endswith('.' + d) for d in TWITTER_DOMAINS)
    except Exception:
        return False

def parse(url):
    try:
        parsed = urllib.parse.urlparse(url)
        match_result = re.search(r'/status(?:es)?/(\d+)', parsed.path)
        if not match_result:
            return None
            
        tweet_id = match_result.group(1)
        api_url = f"https://api.fxtwitter.com/2/status/{tweet_id}"
        
        headers = {'User-Agent': 'Mozilla/5.0'}
        response = requests.get(api_url, headers=headers, timeout=5)
        if not response.ok:
            return None
            
        data = response.json()
        
        if data.get('code') == 200 and data.get('status'):
            status = data['status']
            author = status.get('author', {}).get('name', 'Unknown')
            text = status.get('text', '')
            reposts = status.get('reposts', 0)
            likes = status.get('likes', 0)
            
            media_info = []
            if status.get('media'):
                media = status['media']
                if media.get('photos') and isinstance(media['photos'], list) and len(media['photos']) > 0:
                    media_info.append(f"{len(media['photos'])} 張圖片")
                if media.get('video') or media.get('videos'):
                    media_info.append("包含影片")
                    
            media_text = f" [附帶: {', '.join(media_info)}]" if media_info else ""
            
            return f"[Twitter 推文 | 作者: {author} | 轉推: {reposts} 喜歡: {likes}{media_text}]\n{text}"
    except Exception as e:
        logger.warning(f"[Twitter 網址解析失敗] {str(e)}")
        
    return None

File: src/utils/urlMetadata/twitterParser.js
const URL = require('url');

const TWITTER_DOMAINS = ['twitter.com', 'x.com', 'fxtwitter.com', 'fixupx.com', 'vxtwitter.com'];

/**
 * 檢查是否為 Twitter/X 相關網址
 */
function match(urlString) {
    try {
        const parsed = new URL.URL(urlString);
        return TWITTER_DOMAINS.includes(parsed.hostname) || TWITTER_DOMAINS.some(d => parsed.hostname.endsWith('.' + d));
    } catch (e) {
        return false;
    }
}

/**
 * 解析 Twitter/X 網址並提取內容
 */
async function parse(urlString) {
    try {
        const parsed = new URL.URL(urlString);
        // 匹配推文 ID： /status/:id 或 /statuses/:id
        const matchResult = parsed.pathname.match(/\/status(?:es)?\/(\d+)/);
        if (!matchResult) return null;
        
        const id = matchResult[1];
        const apiUrl = `https://api.fxtwitter.com/2/status/${id}`;
        
        const response = await fetch(apiUrl);
        if (!response.ok) return null;
        
        const data = await response.json();
        
        if (data.code === 200 && data.status) {
            const author = data.status.author ? data.status.author.name : 'Unknown';
            const text = data.status.text || '';
            const reposts = data.status.reposts || 0;
            const likes = data.status.likes || 0;
            
            // 媒體內容判斷
            let mediaInfo = [];
            let images = [];
            
            if (data.status.media) {
                if (data.status.media.photos && Array.isArray(data.status.media.photos) && data.status.media.photos.length > 0) {
                    mediaInfo.push(`${data.status.media.photos.length} 張圖片`);
                    images = data.status.media.photos.map(p => p.url).filter(Boolean);
                }
                if (data.status.media.video || data.status.media.videos) {
                    mediaInfo.push(`包含影片`);
                }
            }
            const mediaText = mediaInfo.length > 0 ? ` [附帶: ${mediaInfo.join(', ')}]` : '';

            const resultText = [
                `[Twitter 推文 | 作者: ${author} | 轉推: ${reposts} 喜歡: ${likes}${mediaText}]`,
                `${text}`
            ].join('\n');

            return {
                text: resultText,
                images: images
            };
        }
    } catch (e) {
        console.error(`[Twitter 網址解析失敗] ${e.message}`);
    }
    
    // 如果解析失敗但仍是 Twitter 網址，回傳 null 讓其他 parser 接手
    return null;
}

module.exports = {
    match,
    parse
};
